src/main/java/uk/nhs/digital/nhsconnect/nhais/mesh/MeshService.java

```diff
@@ -22,57 +23,82 @@
+        this.pollingCycleDurationInSeconds = pollingCycleDurationInSeconds;
     }
 
-    @Scheduled(fixedRateString = "${nhais.mesh.scanMailboxIntervalInMilliseconds}")
+    @Scheduled(fixedRateString = "${nhais.mesh.wakeupIntervalInMilliseconds}")
     public void scanMeshInboxForMessages() {
         if (!meshMailBoxScheduler.isEnabled()){
             LOGGER.warn("Not running the MESH mailbox polling cycle because it is disabled. Set variable " +
                 "NHAIS_SCHEDULER_ENABLED to true to enable it.");
             return;
         }
         LOGGER.info("Requesting lock from database to run MESH mailbox polling cycle");
-        if (meshMailBoxScheduler.hasTimePassed(scanDelayInSeconds)) {
-            LOGGER.info("Starting MESH mailbox polling cycle");
-            meshClient.authenticate();
-            List<String> inboxMessageIds = meshClient.getInboxMessageIds();
-            LOGGER.info("There are {} messages in the MESH mailbox", inboxMessageIds.size());
-            for (String messageId : inboxMessageIds) {
-                try {
-                    LOGGER.debug("Downloading message id {}", messageId);
-                    InboundMeshMessage meshMessage = meshClient.getEdifactMessage(messageId);
-                    LOGGER.debug("Publishing content of message id {} to inbound mesh MQ", messageId);
-                    inboundQueueService.publish(meshMessage);
-                    LOGGER.debug("Acknowledging message id {} on MESH API", messageId);
-                    meshClient.acknowledgeMessage(meshMessage.getMeshMessageId());
-                } catch (MeshWorkflowUnknownException ex) {
-                    LOGGER.warn("Message id {} has an unsupported workflow id {} and has been left in the inbox.", messageId, ex.getWorkflowId());
-                } catch (Exception ex) {
-                    LOGGER.error("Error during reading of MESH message. Message id: {}", messageId, ex);
-                    //ignore exception and try to download next message
+        StopWatch pollingCycleElapsedTime = new StopWatch();
+        pollingCycleElapsedTime.start();
+        if (meshMailBoxScheduler.hasTimePassed(pollingCycleMinimumIntervalInSeconds)) {
+            List<String> inboxMessageIds = authenticateAndGetInboxMessageIds();
+            for (int i = 0; i < inboxMessageIds.size(); i++) {
+                String messageId = inboxMessageIds.get(i);
+                if(sufficientTimeRemainsInPollingCycle(pollingCycleElapsedTime)) {
+                    processSingleMessage(messageId);
+                } else {
+                    LOGGER.warn("Insufficient time remains to complete the polling cycle. Processed {} of {} messages from inbox.", i + 1, inboxMessageIds.size());
+                    break;
                 }
             }
-            LOGGER.info("Completed MESH mailbox polling cycle");
+            LOGGER.info("Completed MESH mailbox polling cycle. Processed all messages from inbox.");
```

Comment:
Wouldn't this line execute even when not all messages from inbox were processed?

Reply:
Correct - I have changed the `break` to a `return` so that this is only execute if the loop completes w/o short-circuit.